src/adapter/src/catalog/inner.rs

```diff
@@ -0,0 +1,158 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Functionality belonging to the catalog but extracted to control file size.
+
+use std::collections::BTreeSet;
+
+use mz_audit_log::{EventDetails, EventType, VersionedEvent};
+use mz_controller::clusters::ClusterId;
+use mz_ore::collections::CollectionExt;
+use mz_repr::{GlobalId, Timestamp};
+use mz_sql::catalog::CatalogItem as SqlCatalogItem;
+use mz_sql_parser::ast::display::AstDisplay;
+use mz_sql_parser::ast::{Ident, RawClusterName, Statement};
+use mz_storage_client::types::sources::IngestionDescription;
+
+use crate::catalog::storage::Transaction;
+use crate::catalog::{
+    catalog_type_to_audit_object_type, BuiltinTableUpdate, Catalog, CatalogEntry, CatalogItem,
+    CatalogState, DataSourceDesc, Error, ErrorKind, MaterializedView, Sink, Source,
+};
+use crate::session::Session;
+use crate::AdapterError;
+
+impl Catalog {
+    /// Update catalog in response to an alter set in cluster operation.
+    pub(super) fn transact_alter_set_in_cluster(
+        state: &mut CatalogState,
+        tx: &mut Transaction,
+        builtin_table_updates: &mut Vec<BuiltinTableUpdate>,
+        oracle_write_ts: Timestamp,
+        drop_ids: &BTreeSet<GlobalId>,
+        audit_events: &mut Vec<VersionedEvent>,
+        session: Option<&Session>,
+        id: GlobalId,
+        cluster_id: ClusterId,
+    ) -> Result<(), AdapterError> {
+        let entry = state.get_entry(&id);
+        let name = entry.name().clone();
+
+        let cluster = state.get_cluster(cluster_id);
+
+        // Prevent changing system objects.
+        if entry.id().is_system() {
+            let schema_name = state
+                .resolve_full_name(&name, session.map(|session| session.conn_id()))
+                .schema;
+            return Err(AdapterError::Catalog(Error::new(
+                ErrorKind::ReadOnlySystemSchema(schema_name),
+            )));
+        }
+
+        // Only internal users can move objects to system clusters
+        if cluster.id.is_system()
+            && !session
+                .map(|session| session.user().is_internal())
+                .unwrap_or(false)
+        {
+            return Err(AdapterError::Catalog(Error::new(
+                ErrorKind::ReadOnlyCluster(cluster.name.clone()),
+            )));
+        }
+
+        // Since the catalog serializes the items using only their creation statement
+        // and context, we need to parse and rewrite the with options in that statement.
+        // (And then make any other changes to the object definition to match.)
+        let mut stmt = mz_sql::parse::parse(entry.create_sql())
+            // TODO: Should never happen
+            .expect("invalid create sql persisted to catalog")
+            .into_element()
+            .ast;
+
+        // Patch AST.
+        let stmt_in_cluster = match &mut stmt {
```

Comment:
Need Index here too?

Reply:
Added, although it's not clear we'll ever implement it. But it can avoid surprises down the road!